src/utils/cycleTracker.ts

```diff
@@ -0,0 +1,112 @@
+import * as fs from 'fs'
+import * as path from 'path'
+import * as Logger from '../Logger'
+import { getCheckpointStatusesByUnifiedStatus } from '../dbstore/checkpointStatus'
+
+interface CycleTrackerData {
+  lastUpdatedCycle: number
+  lastUpdatedTimestamp: number
+}
+
+const CYCLE_TRACKER_FILE = path.join(process.cwd(), 'cycle-tracker.json')
+
+/**
+ * Gets the last updated cycle from the tracker file
+ * @returns The last updated cycle number, or 0 if not found
+ */
+export function getLastUpdatedCycle(): number {
+  try {
+    // If the file does not exist, create it with default values
+    if (!fs.existsSync(CYCLE_TRACKER_FILE)) {
+      const trackerData: CycleTrackerData = {
+        lastUpdatedCycle: 0,
+        lastUpdatedTimestamp: 0
+      }
+
+      fs.writeFileSync(CYCLE_TRACKER_FILE, JSON.stringify(trackerData, null, 2), 'utf8')
+    }
+
+    const data = fs.readFileSync(CYCLE_TRACKER_FILE, 'utf8')
+    const trackerData: CycleTrackerData = JSON.parse(data)
+    return trackerData.lastUpdatedCycle
+  } catch (error) {
+    Logger.mainLogger.error('Error reading cycle tracker file:', error)
+    return 0
+  }
+}
+
+/**
+ * Updates the last updated cycle in the tracker file
+ * @param cycle The cycle number to update
+ */
+export function updateLastUpdatedCycle(cycle: number): void {
+  const lastUpdatedCycle = getLastUpdatedCycle()
+  if (cycle > lastUpdatedCycle) {
+    try {
+      const trackerData: CycleTrackerData = {
+        lastUpdatedCycle: cycle,
+        lastUpdatedTimestamp: Date.now()
+      }
+
+      fs.writeFileSync(CYCLE_TRACKER_FILE, JSON.stringify(trackerData, null, 2), 'utf8')
+      Logger.mainLogger.debug(`Updated cycle tracker to cycle ${cycle}`)
+    } catch (error) {
+      Logger.mainLogger.error('Error updating cycle tracker file:', error)
+    }
+  }
+}
+
+/**
+ * Gets the latest unified cycle counter value from the checkpoint status database
+ * excluding the latest 21 cycles since the checkpoint system works on those
+ * @returns The latest unified cycle number before the latest 21 cycles, or 0 if not found
+ */
+async function getLatestUnifiedCycle(): Promise<number> {
+  try {
+    // Get all checkpoint statuses with unified status = true
+    const unifiedStatuses = await getCheckpointStatusesByUnifiedStatus(true)
+    
+    if (!unifiedStatuses || unifiedStatuses.length === 0) {
+      Logger.mainLogger.warn('No unified cycle statuses found')
+      return 0
+    }
+    
+    // Sort by cycle in descending order
+    const sortedStatuses = unifiedStatuses.sort((a, b) => b.cycle - a.cycle)
+    
+    // Skip the latest 21 cycles (if available) and get the next unified cycle
+    const LATEST_CYCLES_TO_SKIP = 21
```

Comment:
There should be some existing constant for this value 21 (basically the number of cycles that checkpoints systems keeps in memory at a time)